### PR TITLE
MUMMNG-2675 : Health status per key

### PR DESCRIPTION
**Pull request by timlevett**
First step into creating a health check for this rest proxy for each key. Decided to go simple and just have an in memory map based of key. Each object has a map that is the return code and epoch time of the last time it got that code.

Example response:

``` json
{
   "uwmadisonreddit":{
     "key": "uwmadisonreddit",
     "lastResultMap":{
       "429": 1473713497451
     }
   }
}
```

### CODE REVIEW

**Comment by Doug-Reed** [comment]
:+1:

File: build.gradle [hunk 1]
```diff
@@ -10,7 +10,7 @@ plugins {
 subprojects {
 
     group = 'edu.wisc.my.restproxy'
-    version = '3.0.1'
+    version = '3.0.3'
```
Comment (apetro) [inline build.gradle]: Pedantic: adds a new feature, so SemVer as I understand it would prescribe `3.1.0`.

Comment (timlevett) [reply build.gradle]: Sure, that seems fine.

File: rest-proxy-core/src/main/groovy/edu/wisc/my/restproxy/model/ProxyReport.groovy [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+@Builder(builderStrategy = SimpleStrategy)
+@Canonical
+@CompileStatic
+public class ProxyReport {
+  public ProxyReport(String key){
+    this.key = key;
+  }
+  String key;
+  Map<String, Long> lastResultMap = new HashMap<>();
```
Comment (apetro) [inline rest-proxy-core/src/main/groovy/edu/wisc/my/restproxy/model/ProxyReport.groovy:20]: I'd prefer a comment here with excruciating clarity about what the keys and values of this map represent.

**Comment by apetro** [comment]
:+1:

Thoughts:
1. Reading about Clojure and in particular [this Rich Hickey talk](https://youtu.be/VSdnJDO-xdg?t=22m13s) makes me interested in the pros/cons of the strongly typed Interface and Implementation approach vs treating data as just data, using a Map, and calling that good enough until it proves inadequate.  That is, this changeset represents good practices traditional Java / Groovy code. There's a `ReportService` interface and its implementation and a `ProxyReport` object. A traditional Java-ist would say A+, and would be right to say so. But all of that is just a `Map` or maybe `Map`s or `Map`s of `Map`s. A Clojurist might go that route, and might not be wrong either. Interesting to think about whether the code would be more or less "good" going that route, and what "good" even means.
2. I love the simplest thing that could possibly work initial in-memory `Map` implementation. Not only is this good enough for now, it's likely to be quite awesome maybe for a long time. Nonetheless, interesting to look around a bit and see whether and what `Spring` or so has to offer for health monitoring use cases. I haven't looked, but this changeset did give me that itchy look-around-for-a-library-for-this feeling.

EDIT: Had previously, confusingly, linked the wrong Rich Hickey talk.

**Pull request merged**